### ISSUE TITLE
No support for 'links' key, which optionally contains 'self', 'next', 'last' and 'related'

### ISSUE THREAD

**Issue by djones** [Netflix/fast_jsonapi#94]
Part of the JSON API is to support [links](http://jsonapi.org/format/#document-links) but it doesn't appear like there's any way to achieve this right now.

Sample output of this:

```json
{
  "links": {
    "self": "http://example.com/articles",
    "next": "http://example.com/articles?page[offset]=2",
    "last": "http://example.com/articles?page[offset]=10"
  },
  "data": [{
    "type": "articles",
    "id": "1",
    "attributes": {
      "title": "JSON API paints my bikeshed!"
    }
  }]
}
```

**Comment by laertispappas** [comment]
@shishirmk @djones Is anyone working on this? Can I give it a try?

**Comment by shishirmk** [comment]
@laertispappas go for it. Give it a try. Just make sure you branch of `dev`

**Comment by djones** [comment]
I'm not working on it @laertispappas. Would ❤️❤️❤️ to see this shipped though.

**Comment by laertispappas** [comment]
Great then will prepare one PR this weekend. Thanks :)

**Comment by shishirmk** [comment]
@djones @laertispappas links support is merged into `dev` branch. Please give it a try and let me know if it works for you too.

**Comment by djones** [comment]
Thanks for the ping @shishirmk. I'll try it out today and report back.

**Comment by djones** [comment]
This is good progress, thanks for the PR @onumossn.

#117 allows you to have `links` outside of `data`. This is often used for pagination. But it's missing the ability to add `links` inside `data`. Here's what I mean...

![links](https://user-images.githubusercontent.com/2478/37689241-14a0d7e2-2c61-11e8-84d4-45b1d7f1aa69.png)

(the example JSON is pulled from the http://jsonapi.org homepage)

So I think this issue needs to remain open until that bit is supported too.

**Comment by shishirmk** [comment]
Thanks for your response @djones how do you specify those links today? Can you share a sample serializer?

**Comment by djones** [comment]
Sure!

```ruby
class A::BaseSerializer
  include JSONAPI::Serializer

  def base_url
    '/api/v1'
  end
end
```

```ruby
class A::ASerializer < A::BaseSerializer
  attribute :x
  attribute :y

  has_many :x

  def self_link
    "#{base_url}/x"
  end

  def relationship_self_link(_attribute_name)
    nil
  end

  def relationship_related_link(attribute_name)
    "#{base_url}/x/#{attribute_name}"
  end
end
```

We're currently using https://github.com/fotinakis/jsonapi-serializers

There's some docs under customization https://github.com/fotinakis/jsonapi-serializers#more-customizations about `self_link`, `relationship_self_link` and `relationship_related_link`.

**Comment by jshow** [comment]
`links` inside `data` is presently blocking me.

Anybody working on this ? else I'll take a look

**Comment by jshow** [comment]
How does this look ? Links within Data

https://github.com/Netflix/fast_jsonapi/pull/161

**Comment by djones** [comment]
Neat 👍

**Comment by anirudh-dasu** [comment]
Any ETA on merging #161 ?

**Comment by shishirmk** [comment]
Added my comments about #161 needs some work as it mixes two features.

**Comment by shishirmk** [comment]
This should be released in version 1.2 of the gem
